File: dottorato/bplnn.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
# Copyright (C) 2008 Matteo Bertini

import os
import sys
import math
import time
import pprint

import numpy as np
from numpy import matlib
from numpy import random
import pylab

VERBOSE = 1

if VERBOSE > 1:
    np.seterr(over="raise")
def debug(message):
    if VERBOSE > 1:
        print "DEBUG:", message
def info(message):
    if VERBOSE > 0:
        print "INFO:", message

def print_exc_plus():
    """
    Print the usual traceback information, followed by a listing of all the
    local variables in each frame.
    """
    import traceback
    tb = sys.exc_info()[2]
    while 1:
        if not tb.tb_next:
            break
        tb = tb.tb_next
    stack = []
    f = tb.tb_frame
    while f:
        stack.append(f)
        f = f.f_back
    stack.reverse()
    traceback.print_exc()
    print "Locals by frame, innermost last"
    for frame in stack:
        print
        print "Frame %s in %s at line %s" % (frame.f_code.co_name,
                                             frame.f_code.co_filename,
                                             frame.f_lineno)
        for key, value in frame.f_locals.items():
            print "\t%20s = " % key,
            #We have to be careful not to cause a new error in our error
            #printer! Calling str() on an unknown object could cause an
            #error we don't want.
            try:
                print value
            except:
                print "<ERROR WHILE PRINTING VALUE>"

def timed(func):
    start_time = time.time()
    try:
        func()
    except:
        print_exc_plus()
    print "Time:", (time.time() - start_time)

def stats(inputs, targets):
    return "\n".join(["patterns = %s" % len(inputs),
                      "features = %s" % len(inputs[0]),
                      "targets  = %s" % len(targets[0])])


def trace(s, sep='-'):
    s = (" %s " % s).center(70, sep)
    row = sep*len(s)
    info("\n".join([row, s, row]))


def assertEqual(a, b, message=None):
    """
    Verbose assertEqual
    """
    if not message:
        message = "Was supposed %s == %s" % (a, b)
    if a != b:
        raise ValueError(message)

def sigmoid(v):
    return 1.0 / (1.0 + np.exp(-v))
def sigmoid_deriv(v):
    return np.multiply(v, (1.0 - v))
def softmax(v):
    vv = np.exp(v - np.max(v, axis=1))
    return vv / np.sum(vv, axis=1)

class LinearSumOfSquares:
    @staticmethod
    def delta_outputs(layer):
        #debug("targets = %s\noutputs = %s" % (layer.targets, layer.outputs))
        return (layer.targets - layer.outputs)
    @staticmethod
    def outputs(layer):
        return layer.activations
    @staticmethod
    def delta_inputs(layer):
        #debug("inputs = %s\ndelta_outputs = %s\nweights.T = %s" % (layer.inputs, layer.delta_outputs, layer.weights.T))
        return (layer.delta_outputs * layer.weights.T)
    @staticmethod
    def errors(layer):
        return np.sum(np.mat(np.array(layer.targets - layer.outputs)**2), axis=1)
    @staticmethod
    def delta_weights(layer):
        #debug("inputs.T = %s\ndelta_outputs = %s" % (layer.inputs.T, layer.delta_outputs))
        return (layer.inputs.T * layer.delta_outputs)

class SigmoidSumOfSquares:
    @staticmethod
    def delta_outputs(layer):
        #debug("targets = %s\noutputs = %s" % (layer.targets, layer.outputs))
        return np.multiply(sigmoid_deriv(layer.outputs), (layer.targets - layer.outputs))
    @staticmethod
    def outputs(layer):
        return sigmoid(layer.activations)
    @staticmethod
    def delta_inputs(layer):
        #debug("inputs = %s\ndelta_outputs = %s\nweights.T = %s" % (layer.inputs, layer.delta_outputs, layer.weights.T))
        return np.multiply(sigmoid_deriv(layer.inputs), (layer.delta_outputs * layer.weights.T))
    @staticmethod
    def errors(layer):
        return np.sum(np.mat(np.array(layer.targets - layer.outputs)**2), axis=1)
    @staticmethod
    def delta_weights(layer):
        #debug("inputs.T = %s\ndelta_outputs = %s" % (layer.inputs.T, layer.delta_outputs))
        return (layer.inputs.T * layer.delta_outputs)

class SoftmaxCrossEntropy:
    @staticmethod
    def delta_outputs(layer):
        return layer.outputs - layer.targets # -(layer.targets - layer.outputs)
    @staticmethod
    def outputs(layer):
        return softmax(layer.activations)
    @staticmethod
    def delta_inputs(layer):
        return (layer.delta_outputs * layer.weights.T)
    @staticmethod
    def errors(layer):
        return -np.sum(np.multiply(layer.targets, np.log(layer.outputs)), axis=1)
    @staticmethod
    def delta_weights(layer):
        #debug("inputs.T = %s\ndelta_outputs = %s" % (layer.inputs.T, layer.delta_outputs))
        return (layer.inputs.T * layer.delta_outputs)


class Layer:
    def __init__(self, n_in, n_out, squash=SigmoidSumOfSquares):
        self.n_in = n_in
        self.n_out = n_out
        self.squash = squash
        self.inputs = None
        self.delta_inputs = None
        self.weights = np.mat(np.random.randn(n_in, n_out)*1.0)
        self.activations = None
        self.outputs = None
        self.delta_outputs = None
        self.targets = None
        self.errors = None
    def propagate(self, inputs):
        self.inputs = inputs
        #debug("inputs = %s\nweights = %s" % (self.inputs, self.weights))
        self.activations = self.inputs*self.weights
        self.outputs = self.squash.outputs(self)
        #debug("outputs = %s" % self.outputs)
        return self.outputs
    def backPropagate(self, targets=None, delta_outputs=None):
        if targets != None:
            self.targets = targets
            self.delta_outputs = self.squash.delta_outputs(self)
            #debug("delta_outputs = %s" % self.delta_outputs)
            self.errors = self.squash.errors(self)
            #debug("errors = %s" % self.errors)
        elif delta_outputs != None:
            self.delta_outputs = delta_outputs
            #debug("delta_outputs = %s" % self.delta_outputs)
        else:
            raise ValueError("provide 'targets' or 'delta_outputs'")
        self.delta_inputs = self.squash.delta_inputs(self)
        #debug("delta_inputs = %s\nerrors = %s" % (self.delta_inputs, self.errors))
        return self.delta_inputs
    def updateWeights(self, learn):
        self.weights += learn * self.squash.delta_weights(self)
        #debug("weights = %s" % (self.weights))
        return self.weights
    def __repr__(self):
        return "<Layer %d %d %s>" % (self.n_in, self.n_out, self.squash.__name__)
        
        
class AbstractNetwork:
    def train(self, inputs, targets, iterations=1000, learn=0.05):
        count = iterations
        step = 10 + int(math.sqrt(iterations))
        num = inputs.shape[0]
        batch_size = 1 + num / 20 # numbatches... tune it for you system/dataset
        info((" TRAIN batch_size=%s " % batch_size).center(70, "#"))
        ind = np.arange(num)
        weights_before = []
        for c, layer in enumerate(self.layers):
            weights_before.append(layer.weights.copy())
            pylab.imshow(layer.weights, cmap=pylab.cm.gray)
            pylab.savefig("%s (%1d) before" % (self, c))
        while count:
            count -= 1
            error = 0.0
            np.random.shuffle(ind)
            for batch in range(0, num, batch_size):
                selection = ind[batch:batch + batch_size]
                self.propagate(inputs[selection])
                self.backPropagate(targets[selection])
                error += self.errors.sum()
                self.updateWeights(learn)
            if not count % step:
                info("iter(%s) error = %s" % (count, error))
        for c, layer in enumerate(self.layers):
            pylab.imshow(layer.weights, cmap=pylab.cm.gray)
            pylab.savefig("%s (%1d) post" % (self, c))
    def test(self, inputs, targets):
        info(" TEST ".center(70, "#"))
        def cls(arg):
            if arg.shape[1] == 1:
                return int(arg > 0.5)
            else:
                return arg.argmax()
        #for i in range(inputs.shape[0])[:10]:
        #    info("%s -> %s (%s)" % (inputs[i], cls(self.propagate(inputs[i])), cls(targets[i])))
        res = {}
        for i in range(inputs.shape[0]):
            output = cls(self.propagate(inputs[i]))
            target = cls(targets[i])
            res.setdefault(target, {True: 0.0, False: 0.0})
            res[target][target == output] += 1.0
            res[target]["acc"] = 100 * res[target][True] / (res[target][True] + res[target][False])
        info(pprint.pformat(res))
        info("mean acc = %f" % np.mean([res[t]["acc"] for t in res]))


class ShallowNetwork(AbstractNetwork):
    def __init__(self, n_in, n_hid, n_out, bias=True, squash=SigmoidSumOfSquares):
        self.bias = bias
        self.n_nodes = [n_in, n_hid, n_out]
        if bias:
            n_in = n_in + 1
        random.seed(123)
        self.layers = [Layer(n_in, n_hid), Layer(n_hid, n_out, squash=squash)]
    def propagate(self, inputs):
        if self.bias:
            inputs = np.append(inputs, np.ones((inputs.shape[0],1)), axis=1)
        #debug(" propagate IN ".center(70, "-"))
        hiddens = self.layers[0].propagate(inputs)
        #debug(" propagate OUT ".center(70, "-"))
        outputs = self.layers[1].propagate(hiddens)
        return outputs
    def backPropagate(self, targets):
        #debug(" backPropagate OUT ".center(70, "-"))
        delta_inputs = self.layers[1].backPropagate(targets=targets)
        self.errors = self.layers[1].errors
        #debug(" backPropagate IN ".center(70, "-"))
        return self.layers[0].backPropagate(delta_outputs=delta_inputs)
    def updateWeights(self, learn):
        #debug(" updateWeights IN ".center(70, "-"))
        self.layers[0].updateWeights(learn)
        #debug(" updateWeights OUT ".center(70, "-"))
        self.layers[1].updateWeights(learn)
    def dump(self):
        return {"ShallowNetwork": self.layers}
    def __str__(self):
        return "<ShallowNetwork %s>" % self.n_nodes


class DeepNetwork(AbstractNetwork):
    def __init__(self, n_nodes, bias=True):
        self.bias = bias
        self.n_nodes = n_nodes
        self.prepare_iterations = 1
        if bias:
            n_nodes[0] += 1 # bias
        self.layers = [Layer(n_in, n_out) for n_in, n_out in zip(n_nodes[:-1], n_nodes[1:])]
        self.layers += [Layer(n_nodes[-1], n_nodes[-1])] # descramble autoencoder
    def propagate(self, inputs):
        if self.bias:
            inputs = np.append(inputs, np.ones((inputs.shape[0],1)), axis=1)
        outputs = self.layers[0].propagate(inputs)
        for layer in self.layers[1:]:
            outputs = layer.propagate(outputs)
        return outputs
    def backPropagate(self, targets):
        delta_inputs = self.layers[-1].backPropagate(targets=targets)
        self.errors = self.layers[-1].errors
        for layer in reversed(self.layers[:-1]):
            delta_inputs = layer.backPropagate(delta_outputs=delta_inputs)
        return delta_inputs
    def updateWeights(self, learn):
        for layer in reversed(self.layers):
            if layer != self.layers[-1]:
                learn = learn / len(self.layers)#self.prepare_iterations
            layer.updateWeights(learn)
    def prepare(self, inputs, iterations, learn):
        self.prepare_iterations = iterations
        info(" PREPARE ".center(70, "o"))
        if self.bias:
            inputs = np.append(inputs, np.ones((inputs.shape[0],1)), axis=1)
        iters = map(int, np.linspace(iterations, iterations/2, len(self.layers)))
        for c, layer in enumerate(self.layers[:-1]):
            info((" (%d) %s " % (c, layer)).center(70, "#"))
            auto_net = ShallowNetwork(layer.n_in, layer.n_out, layer.n_in, bias=False)
            auto_net.layers[0].weights = layer.weights
            auto_net.train(inputs, inputs, iters[c])
            layer.weights = auto_net.layers[0].weights
            auto_net.propagate(inputs)
            inputs = auto_net.layers[0].outputs
    def dump(self):
        return {"DeepNetwork": [l.dump() for l in self.layers]}
    def __str__(self):
        return "<DeepNetwork %s>" % self.n_nodes


def demo(iterations=1000, learn=0.05):
    # Teach network XOR function
    patterns = np.mat([[0.0,0.0, 0.0],
                       [0.0,1.0, 1.0],
                       [1.0,0.0, 1.0],
                       [1.0,1.0, 0.0]])

    patterns = np.mat([[0.0,0.0, 0.0,0.0],
                       [0.0,1.0, 0.0,1.0],
                       [1.0,0.0, 1.0,0.0],
                       [1.0,1.0, 1.0,1.0]])
    inputs, targets = patterns[:,:-2], patterns[:,-2:]
    # create a network
    net = ShallowNetwork(2, 5, 2, squash=SoftmaxCrossEntropy)
    #net = DeepNetwork([2, 5, 1])
    # train it with some patterns
    #for i in range(1):
    #    net.prepare(patterns, 50000, 0.05)
    # test it
    #print net.dump()
    #net.test(patterns)
    print net
    if isinstance(net, DeepNetwork):
        net.prepare(inputs, iterations, learn)
    net.train(inputs, targets, iterations, learn)
    net.test(inputs, targets)
    return net


if __name__ == "__main__":
    if __debug__:
        import doctest
        doctest.testmod()
    timed(demo)

File: dottorato/bplnn.ml
(* BackPropagation Layered Neural Network                       *)
(* Copyright (c) 2008 Matteo Bertini                            *)

(* ----------------- Extending Array module ------------------- *)
module Array = struct
  include Array

  let reverse array =
    (**
    Array.reverse [|1;2;3;4|];;
    - : int array = [|4; 3; 2; 1|]
    *)
    let len = length array in
    let rget array n _ = array.(len - n - 1) in
    mapi (rget array) array

  let sum array =
    fold_left (+.) 0.0 array

  let map2 f array1 array2 =
    let len1 = Array.length array1
    and len2 = Array.length array2 in
    if len1 = len2 then
      Array.mapi (fun i array1_i -> (f array1_i array2.(i))) array1
    else
      failwith "incompatible lenghts for map2"

  let dot array1 array2 =
    (**
    let a1 = [|1.0; 2.0; 3.0|];;
    let a2 = [|2.0; 2.0; -1.0|];;
    Array.dot a1 a2;;
    - : float = 3.
    let a3 = [|2.0; 2.0;|];;
    Array.dot a1 a3;;
		Exception: Failure "incompatible lenghts for map2".
    *)
    sum (map2 ( *. ) array1 array2)
 
  let for_all f array =
    List.for_all f (Array.to_list array)

  let init_matrix rows cols afunc =
    (**
    let randfunc _ _ = Random.rand (-2.0) 2.0;;
    Array.init_matrix 2 3 randfunc;;
    *)
    let matrix = make_matrix rows cols 0.0 in
		for i = 0 to rows - 1 do
			for j = 0 to cols - 1 do
				matrix.(i).(j) <- afunc i j 
			done;
		done;
		matrix

  let transposed_matrix amatrix =
    (**
    let counter =
      let c = ref 0.0 in
      let count () = c := !c +. 1.0; !c in
      count;;
    let count _ _ = counter();;
    let matrix = Array.init_matrix 2 3 count;;
    val matrix : float array array = [|[|1.; 2.; 3.|]; [|4.; 5.; 6.|]|]
    Array.transposed_matrix matrix;;
    - : float array array = [|[|1.; 4.|]; [|2.; 5.|]; [|3.; 6.|]|]
    *)
    let rows = length(amatrix)
    and cols = length(amatrix.(0))
    and get_t r c = amatrix.(c).(r) in
    init_matrix cols rows get_t
end

(* ----------------- Extending Random module ------------------- *)
module Random = struct
  include Random

  let rand min max =
    (**
    Random.rand (-2.0) 2.0;;
    *)
    min +. (max -. min) *. (float 1.0) (* Random.float bound = [0, bound) *)
end

(* ------------------------- Loss --------------------------- *)

module Sigmoid = struct
  (**
  Sigmoid.func 0.5;;
  - : float = 0.622459331201854593
  Sigmoid.deriv 0.5;;
  - : float = 0.25.
  *)
  let func x =
    1.0 /. (1.0 +. exp(-.x))
  let deriv x =
    x *. (1.0 -. x)
	let loss output target =
	  (**
	  qloss 2.0 1.0;;
	  - : float = 2.
	  *)
	  (deriv output) *. (target -. output)
end

(* ------------------------- Layer --------------------------- *)

module Layer = struct
  type layer = {
    mutable inputs: float array;
    mutable delta_inputs: float array;
    weights: float array array;
    outputs: float array;
    delta_outputs: float array;
  }

  let layer inputs delta_inputs weights outputs delta_outputs = {
    inputs = inputs;
    delta_inputs = delta_inputs;
    weights = weights;
    outputs = outputs;
    delta_outputs = delta_outputs;
  }

  let make_layer n_in n_out =
    (**
    Layer.make_layer 2 3;;
    *)
    let inputs = Array.make n_in 1.0 in
    let delta_inputs = Array.make n_in 0.0 in
    let outputs = Array.make n_out 1.0 in
    let delta_outputs = Array.make n_out 0.0 in
    let rand _ _ = Random.rand (- 0.5) 0.5 in
    let weights = Array.init_matrix n_out n_in rand in
    layer inputs delta_inputs weights outputs delta_outputs

  let propagate alayer =
    (**
    let alayer = Layer.make_layer 2 3;;
    Layer.propagate alayer;;
    *)
    for k = 0 to Array.length(alayer.outputs) - 1 do
      alayer.outputs.(k) <- (Array.dot alayer.weights.(k) alayer.inputs)
    done;
    alayer

  let propagate_inputs alayer inputs =
    (**
    let alayer = Layer.make_layer 3 3;;
    Layer.propagate_inputs alayer [|2.0; 2.0|];;
    *)
    for j = 0 to Array.length(inputs) - 1 do (* bias is the last in alayer.inputs *)
      alayer.inputs.(j) <- inputs.(j)
    done;
    propagate alayer

  let back_propagate alayer =
    (**
    let alayer = Layer.make_layer 2 3;;
    Layer.back_propagate alayer;;
    *)
    let _weights = Array.transposed_matrix alayer.weights in
    let delta_func j =
      Sigmoid.deriv(alayer.inputs.(j)) *. (Array.dot _weights.(j) alayer.delta_outputs) in
    for j = 0 to Array.length(alayer.inputs) - 1 do
      alayer.delta_inputs.(j) <- delta_func j
    done;
    alayer

  let back_propagate_targets alayer targets =
    (**
    let alayer = Layer.make_layer 2 3;;
    Layer.propagate_inputs alayer [|2.0; 2.0|];;
    Layer.back_propagate_targets alayer [|2.0; 2.0; 2.0|];;
    *)
    for k = 0 to Array.length(targets) - 1 do
      alayer.delta_outputs.(k) <- Sigmoid.loss alayer.outputs.(k) targets.(k)
    done;
    back_propagate alayer

  let update_weights alayer learn =
    (**
    let alayer = Layer.make_layer 2 3;;
    Layer.propagate_inputs alayer [|2.0; 2.0|];;
    Layer.back_propagate_targets alayer [|2.0; 2.0; 2.0|];;
    Layer.update_weights alayer 0.05;;
    *)
    let cumulate_weight k j value =
      alayer.weights.(k).(j) <- alayer.weights.(k).(j) +. value
		and n_in = Array.length(alayer.inputs)
		and n_out = Array.length(alayer.outputs) in
	  for j = 0 to n_in - 1 do
	    for k = 0 to n_out - 1 do
	      cumulate_weight k j (learn *. alayer.delta_outputs.(k) *. alayer.inputs.(j))
	    done;
	  done;
	  alayer

  let sq2error alayer targets =
    (**
    let alayer = Layer.make_layer 2 1;;
    Layer.propagate_inputs alayer [|1.0; 1.0|];;
    Layer.sq2error alayer [|0.0|];;
    *)
    let sum_of_squares array = 
      Array.sum (Array.map (fun x -> x**2.0) array) in
    0.5 *. sum_of_squares (Array.map2 (-.) alayer.outputs targets)

  let connect curr_layer next_layer =
    (**
    let in_layer = Layer.make_layer 2 3;;
    let out_layer = Layer.make_layer 3 1;;
    Layer.connect in_layer out_layer;;
    Layer.propagate_inputs in_layer [|1.0; 1.0|];;
    Layer.propagate out_layer;;
    Layer.back_propagate_targets out_layer [|0.0|];;
    Layer.back_propagate in_layer;;
    *)
    next_layer.inputs <- curr_layer.outputs;
    next_layer.delta_inputs <- curr_layer.delta_outputs;
    [| curr_layer; next_layer |]

end

(* --------------------- ShallowNetwork ----------------------- *)

module ShallowNetwork = struct
  type network = {
    in_layer: Layer.layer;
    out_layer: Layer.layer;
  }
  
  let network in_layer out_layer = {
    in_layer = in_layer;
    out_layer = out_layer;
  }

  let make_network n_in n_hid n_out =
    (**
    let anet = ShallowNetwork.make_network 2 3 1;;
    *)
    let in_layer = Layer.make_layer (n_in + 1) n_hid (* bias *)
    and out_layer = Layer.make_layer n_hid n_out in
    let _ = Layer.connect in_layer out_layer in
    network in_layer out_layer

  let propagate anetwork inputs =
    let _ = Layer.propagate_inputs anetwork.in_layer inputs in
    let _ = Layer.propagate anetwork.out_layer in
    anetwork

  let back_propagate anetwork targets =
    let _ = Layer.back_propagate_targets anetwork.out_layer targets in
    let _ = Layer.back_propagate anetwork.in_layer in
    anetwork

  let update_weights anetwork learn =
    let _ = Layer.update_weights anetwork.in_layer learn in
    let _ = Layer.update_weights anetwork.out_layer learn in
    anetwork

  let sq2error anetwork targets =
    (**
    let anet = ShallowNetwork.make_network 2 3 1;;
    ShallowNetwork.propagate anet [|1.0; 1.0|];;
    ShallowNetwork.back_propagate anet [|0.0|];;
    ShallowNetwork.update_weights anet 0.05;;
    ShallowNetwork.sq2error anet [|0.0|];;
    *)
    Layer.sq2error anetwork.out_layer targets

  let train anetwork inputs targets iterations learn =
    (**
    let anet = ShallowNetwork.make_network 2 5 1;;
    let inputs = [|
                  [|1.0; 1.0|];
                  [|1.0; 0.0|];
                  [|0.0; 1.0|];
                  [|0.0; 0.0|];
                 |];;
    let targets = [|
                   [|0.0|];
                   [|1.0|];
                   [|1.0|];
                   [|0.0|];
                  |];;
    ShallowNetwork.train anet inputs targets 10 0.05;;
    *)
    for count = 0 to iterations do
      let error = ref 0.0 in
			let print = (count mod 1) == 0 in
      if print then
        Printf.printf "iter(%d)\n" count;
      for i = 0 to Array.length(inputs) - 1 do
        let _ = propagate anetwork inputs.(i) in
        let _ = back_propagate anetwork targets.(i) in
        let _ = update_weights anetwork learn in
				let item_error = sq2error anetwork targets.(i) in
        error := !error +. item_error;
	      if print then
	        Printf.printf "  item(%d) item_error: %f\n" i item_error;
      done;
      if print then
        Printf.printf "iter(%d) error: %f\n" count !error;
    done;
    anetwork

end

let _ =
    let anet = ShallowNetwork.make_network 2 5 1
    and inputs = [|
                  [|1.0; 1.0|];
                  [|1.0; 0.0|];
                  [|0.0; 1.0|];
                  [|0.0; 0.0|];
                 |]
    and targets = [|
                   [|0.0|];
                   [|1.0|];
                   [|1.0|];
                   [|0.0|];
                  |] in
    ShallowNetwork.train anet inputs targets 600 0.05
;;

(* vim:cindent sw=2 sts=2 ts=2 et fdm=marker
 *)
